gemini/sample-apps/llamaindex-rag/common/utils.py

```diff
@@ -0,0 +1,248 @@
+    # too large to hold in memory easily.
+    # max_results=1000
+
+    from google.cloud.storage import transfer_manager
+
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(bucket_name)
+
+    blob_names = [
+        blob.name
+        for blob in bucket.list_blobs(
+            prefix=prefix, delimiter=delimiter, max_results=max_results
+        )
+    ]
+
+    results = transfer_manager.download_many_to_path(
+        bucket,
+        blob_names,
+        destination_directory=destination_directory,
+        max_workers=workers,
+    )
+
+    for name, result in zip(blob_names, results):
+        # The results list is either `None` or an exception for each blob in
+        # the input list, in order.
+
+        if isinstance(result, Exception):
+            logger.info(
+                "Failed to download {} due to exception: {}".format(name, result)
+            )
+        else:
+            logger.info(
+                "Downloaded {} to {}.".format(name, destination_directory + name)
+            )
+
+
+def link_nodes(node_list):
+    for i, current_node in enumerate(node_list):
+        if i > 0:  # Not the first node
+            previous_node = node_list[i - 1]
+            current_node.relationships[NodeRelationship.PREVIOUS] = RelatedNodeInfo(
+                node_id=previous_node.node_id
+            )
+
+        if i < len(node_list) - 1:  # Not the last node
+            next_node = node_list[i + 1]
+            current_node.relationships[NodeRelationship.NEXT] = RelatedNodeInfo(
+                node_id=next_node.node_id
+            )
+    return node_list
+
+
+def create_pdf_blob_list(bucket_name, prefix):
+    """
+    Create a list of Blob objects for processing.
+    """
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(bucket_name)
+    blobs = bucket.list_blobs(prefix=prefix)
+    logger.info(blobs)
+    return [
+        Blob(
+            path=f"gs://{bucket_name}/{blob.name}",
+            mimetype=blob.content_type or "application/pdf",
+        )
+        for blob in blobs
+        if blob.name.lower().endswith(".pdf")
+    ]
+
+
+def upload_directory_to_gcs(local_dir_path: str, bucket_name: str, prefix: str):
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(bucket_name)
+
+    for root, dirs, files in os.walk(local_dir_path):
+        for file in files:
+            local_file_path = os.path.join(root, file)
+            relative_path = os.path.relpath(local_file_path, local_dir_path)
+            gcs_blob_name = f"{prefix}/{relative_path}"
+
+            blob = bucket.blob(gcs_blob_name)
+            blob.upload_from_filename(local_file_path)
+            print(f"File {local_file_path} uploaded to {gcs_blob_name}")
+
+
+def ensure_directory_exists(directory):
+    """
+    Ensure that the given directory exists, creating it if necessary.
+    """
+    os.makedirs(directory, exist_ok=True)
```

Comment:
Not sure this needs its own method

Reply:
Fixed in [commit](https://github.com/GoogleCloudPlatform/generative-ai/commit/e0dd2113cfb85f68d8d9ddad31c2317921c52a85)